pagopa/io-dev-api-server
# [IOCOM-437,IOCOM-439,IOCOM-443] Multiple Payments on PN messages

## Short description
This PR adds multiple payments to PN messages and introduces the payment database to later check and retrieve payment statuses.

## List of changes proposed in this pull request
- `src/config`: new PN configuration for new messages with multiply payments, cancelled status and F24 attachments. New flag to switch between the old and new implementation for the payment-requests/:rptId handler implementation (see below)
- `src/feature/pn/payloads/messages`: previous PN related message creation methods have been moved here and new methods for creating multiply payments have been added here
- `src/features/pn/routers/routers`: PN activation router has been moved here
- `src/featureS/pn/services/services`: PN service data have been moved here
- `src/features/pn/types/notificationPaymentInfo`: new data type for PN payment
- `src/features/pn/types/notificationRecipient`: new data type for PN recipient
- `src/payloads/messages`: PN methods have been moved out of here. Methods generating payments have been refactored in order not to throw but to handle errors using Eithers
- `src/payload/response`: delay property has been refactored to a function, in order to be dynamically computed (i.e., now it can change every time a request is made)
- `src/persistence/payment`: new in-memory payment database that allows to store and later retrieve a payment (previously, there was no such features and payments where randomly generated and checked upon payment-check requests)
- `src/populate-persistence`: methods generating messages with payments and remote attachment have been refactored to use pipeline streaming instead of function composition
- `src/routers/features/siciliaVola/secured/index`: router delay switched from property to function
- `src/routers/message`: router delay switched from property to function
- `src/routers/payment`: payment-requests/:rptId handler refactored to either use the legacy system or use the new payment database to retrieve a payment status. A runtime random delay is applied to this function when used with the new payments database
- `src/routers/services_metadata`: removed explicit using of default delay
- `src/routers/walletsV2/configDashboard`: removed explicit using of default delay
- `src/routers/walletsV2/methods/cobadge`: removed explicit using of default delay
- `src/typeS/PaymentStatus`: new types for the payment database, including unpaid, paid and in-error payments, with function methods to use them in pipelines
- `src/types/config`: types for the new PN configuration and payment-requests/:rptId legacy implementation in src/config
- `src/utils/array`: fp-ts extension of the makeBy methods for the Array instance, to allow the single item function generator to return an Either and stop the array creation process
- `src/utils/payment`: added a method to compute an rptId from a PaymentDataWIthRequiredPayee instance
- `src/utils/validator`: validatePayload method split in order to have a new method that returns an Either, if the decode fails, instead of throwing an exception 

## How to test
Run the server instance with differente PN messages configurations. Retrieve and check their data using the `third-party-messages/:id` API and check their payments using the `payment-requests/:rptId` endpoint
